From 1e5193fdba51ca5d862913ce14b27acee04c3ca5 Mon Sep 17 00:00:00 2001
From: Terrence Xu <terrence.xu@intel.com>
Date: Thu, 23 Jul 2015 23:19:24 +0800
Subject: [PATCH 137/403] vgt: initial guest CSB register after gpu reset
 under execlist mode

We must make sure guest always can get latest HW status during ctx
switch, otherwise vgt idle execlist check will return false continuously
until get the right status.
If guest trigger gpu reset, we must initial guest CSB register to let
vgt idle execlist check get the right value.

Corresponding bug number in bugzilla are: #827, #828.

Signed-off-by: Terrence Xu <terrence.xu@intel.com>
---
 drivers/gpu/drm/i915/vgt/execlists.c |    8 ++++++++
 1 file changed, 8 insertions(+)

diff --git a/drivers/gpu/drm/i915/vgt/execlists.c b/drivers/gpu/drm/i915/vgt/execlists.c
index c526777..717f436 100644
--- a/drivers/gpu/drm/i915/vgt/execlists.c
+++ b/drivers/gpu/drm/i915/vgt/execlists.c
@@ -1790,6 +1790,8 @@ void vgt_reset_execlist(struct vgt_device *vgt, unsigned long ring_bitmap)
 {
 	vgt_state_ring_t *rb;
 	int bit, i;
+	uint32_t ctx_ptr_reg;
+	struct ctx_st_ptr_format ctx_ptr_val;
 
 	for_each_set_bit(bit, &ring_bitmap, sizeof(ring_bitmap)) {
 		if (bit >= vgt->pdev->max_engines)
@@ -1809,6 +1811,12 @@ void vgt_reset_execlist(struct vgt_device *vgt, unsigned long ring_bitmap)
 			memset(&rb->execlist_slots[i], 0,
 					sizeof(struct vgt_exec_list));
 
+		ctx_ptr_reg = el_ring_mmio(bit, _EL_OFFSET_STATUS_PTR);
+		ctx_ptr_val.dw = __vreg(vgt, ctx_ptr_reg);
+		ctx_ptr_val.status_buf_write_ptr = DEFAULT_INV_SR_PTR;
+
 		rb->csb_write_ptr = DEFAULT_INV_SR_PTR;
+
+		__vreg(vgt, ctx_ptr_reg) = ctx_ptr_val.dw;
 	}
 }
-- 
1.7.10.4

